torchmultimodal/models/albef.py

```diff
@@ -0,0 +1,244 @@
+            image, text, text_atts
+        )
+        similarity = self._similarity(
+            image_feat, text_feat, image_feat_m, text_feat_m, alpha
+        )
+        image_embeds_neg, text_embeds_neg, text_atts_neg = self._neg_embeddings(
+            image_embeds, text_embeds, text_atts, similarity
+        )
+        vl_embeds = self._multimodal_embeddings(
+            image_embeds,
+            text_embeds,
+            image_embeds_neg,
+            text_embeds_neg,
+            text_atts,
+            text_atts_neg,
+        )
+        return ALBEFOutput(
+            image_embeddings=image_embeds,
+            image_embeddings_m=image_embeds_m,
+            text_embeddings=text_embeds,
+            text_atts=text_atts,
+            vl_embeddings=vl_embeds,
+            similarity=similarity,
+        )
+
+    @torch.no_grad()
+    def _copy_params(self):
+        for model, model_m in [
+            [self.vision_encoder, self.vision_encoder_m],
+            [self.text_encoder, self.text_encoder_m],
+            [self.multimodal_encoder, self.multimodal_encoder_m],
+            [self.vision_proj, self.vision_proj_m],
+            [self.text_proj, self.text_proj_m],
+        ]:
+            for param, param_m in zip(model.parameters(), model_m.parameters()):
+                param_m.data.copy_(param.data)
+                param_m.requires_grad = False
+
+    def _unimodal_embeddings(self, image, text, text_atts):
+        image_embeds = self.vision_encoder(image)
+        text_embeds = self.text_encoder(text, attention_mask=text_atts)
+        image_feat = F.normalize(self.vision_proj(image_embeds[:, 0, :]), dim=-1)
+        text_feat = F.normalize(self.text_proj(text_embeds[:, 0, :]), dim=-1)
+        return image_embeds, text_embeds, image_feat, text_feat
+
+    @torch.no_grad()
+    def _momentum_embeddings(self, image, text, text_atts):
+        self._momentum_update()
+        image_embeds_m = self.vision_encoder_m(image)
+        text_embeds_m = self.text_encoder_m(text, attention_mask=text_atts)
+        image_feat_m = F.normalize(self.vision_proj_m(image_embeds_m[:, 0, :]), dim=-1)
+        text_feat_m = F.normalize(self.text_proj_m(text_embeds_m[:, 0, :]), dim=-1)
+        return image_embeds_m, image_feat_m, text_feat_m
+
+    @torch.no_grad()
+    def _momentum_update(self):
+        for model, model_m in [
+            [self.vision_encoder, self.vision_encoder_m],
+            [self.text_encoder, self.text_encoder_m],
+            [self.multimodal_encoder, self.multimodal_encoder_m],
+            [self.vision_proj, self.vision_proj_m],
+            [self.text_proj, self.text_proj_m],
+        ]:
+            for param, param_m in zip(model.parameters(), model_m.parameters()):
+                param_m.data = param_m.data * self.momentum + param.data * (
+                    1 - self.momentum
+                )
+
+    @torch.no_grad()
+    def _update_queue(self, image_feat_m, text_feat_m):
```

Comment:
need type annotations on all functions - unfortunately our mypy config does not catch this right now